[difftest] draft difftest organization doc (#774)
chipsalliance/t1: difftest/doc/rational.md

#### Objective
The goal of offline difftest is to ensure the correctness of each instruction’s observed behavior during processor execution, while providing a flexible validation mechanism for handling undefined behavior. The validation process involves three steps:

1. Running the driver to generate the DUT (Device Under Test) trace.
2. Using the DUT trace to generate the model trace.
3. Detecting instruction differences by comparing the DUT trace and model trace using the difftest algorithm.

#### Design Choices

##### 1. **DUT Trace Design and Encoding**
   - **Design:** The DUT trace design must consider how to efficiently and accurately record the execution state of each instruction, including the execution address, register values, memory state, etc. Currently, using `printf` to output the simulation address is a preliminary solution, with plans to possibly use XDMA (high-speed data transfer interface) to extract FPGA traces in the future for higher verification efficiency.
   - **Encoding:** Proper encoding of the trace will facilitate fast alignment and parsing in the comparison stage.

##### 2. **Simulator Integration**

>    - **Using SAIL Simulator:** SAIL offers a highly accurate and customizable architecture simulation framework. Despite its slower speed, SAIL’s precision and flexibility make it well-suited for offline verification needs.

Reviewer: Accurate is not appropriate here. Maybe formally declared sounds better

     - **Removing Platform Dependencies:** To simplify the simulation, platform-related parts of SAIL will be removed, focusing on validating the instruction set and processor core.
     - **Function Patch:** For example, in validating FP (floating-point) related instructions, certain functions, such as `fp reduce order`, may need to be replaced.
     - **Handling Undefined Behavior:** To verify undefined behavior, an interface will be provided for the simulator to inject specific behavior. This may involve forking the SAIL project or directly customizing its source code.

##### 3. **DiffTest Algorithm**
   - **Instruction Alignment:** During the comparison process, it’s crucial to ensure that the DUT and model instruction streams are properly aligned. This may require special alignment algorithms to handle challenges arising from out-of-order or parallel execution.

>    - **Result Comparison:** The core of the comparison is to ensure that the execution results of each instruction match. This can be done via streaming-based real-time comparison or through other efficient differential calculation methods.

Reviewer: ```suggestion
   - **Comparison Algorithm:** The core of the comparison algorithm is to ensure that the observance of each instruction being same. Due to most of OoO write-back strategy, comparison commit result is the core reason of difftest framework diverse from different core projects.
```

#### Potential Challenges and Solutions

##### 1. **Memory Behavior Issues**

>    - When memory access order differs, it may lead to processor behavior discrepancies. In such cases, alignment strategies and memory model simulation may be needed to reduce the impact of memory ordering on the behavior.

Reviewer: ```suggestion
   - When memory access order differs, it may lead to processor behavior discrepancies. In such cases, alignment strategies and memory model simulation may be needed.
```

##### 2. **Handling Undefined Behavior**
   - Defining and simulating undefined behavior is a complex task. A flexible interface is needed to allow the simulator to inject undefined behavior when detected, and to record related traces for further analysis.

> #### Conclusion
> The offline difftest system is an innovative validation approach designed to enhance the flexibility and accuracy of verification by decoupling the driver and verification processes, thereby reducing the impact of verification on the driver process. This system not only precisely verifies the results of each instruction but also flexibly handles undefined behavior, offering an efficient and general-purpose validation method for complex processor architectures.

Reviewer: ChatGPT always gives Conclusion, but we are not reach to the conclusion yet.
```suggestion
```
